base/loading.jl

```diff
@@ -1884,8 +1884,12 @@ function _require(pkg::PkgId, env=nothing)
                     @goto load_from_cache
                 end
                 # spawn off a new incremental pre-compile task for recursive `require` calls
-                cachefile = compilecache(pkg, path)
-                if isa(cachefile, Exception)
+                cachefile = maybe_cachefile_lock(pkg, path) do
+                    compilecache(pkg, path)
```

Comment:
It is incorrect to call `compilecache(pkg, path)` here quite yet, because we need to call `@goto load_from_cache` first (now that we have the lock) and prove that the file wasn't created before now. This may require some heavy-duty refactoring though (sorry!).

After this PR, it looks like we also now need to update Pkg to use this new code too, and make sure that each step inside `PKG_PRECOMPILE_HOOK[]` happens behind this exact same lock for `maybe_cachefile_lock(pkg, path)` as well (`PKG_PRECOMPILE_HOOK[]` probably can hold exactly one lock at a time, avoiding any ordering problems that might lead to deadlock).

Reply:
I couldn't see a way to refactor without duplicating the `_require_search_from_serialized` so I just did that for now